Lazy-create thread-specific coverage to immediately process cached edges

evansiroky opened the pull request:
Add one more small improvement that should speed-up calculations during builds that use cached elevations.

evansiroky committed:
Lazy-create thread-specific coverage to immediately process cached edges

evansiroky committed:
Small comment refactor

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java:
```diff
@@ -480,31 +509,15 @@ private void assignMissingElevations(
      * Processes a single street edge, creating and assigning the elevation profile.
      *
      * @param ee the street edge
+     * @param edgeGeomerty The geometry of the edge
```
typo `edgeGeomerty`

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java:
```diff
@@ -241,7 +241,36 @@ public ProcessEdgeTask(StreetWithElevationEdge swee, ConcurrentHashMap<Long, Cov
+                        edgeNeedsProcessing = false;
+                    }
+                }
+            }
```
Extract into its own method.

binh-dam-ibigroup reviewed (changes requested):
Please see extract method request.

evansiroky committed:
Move elevation cache code back into altered processEdge method

evansiroky commented:
Moved some stuff around, hopefully it looks better now.

evansiroky committed:
Refactor a few things

- Use a custom ForkJOinWorkerThreadFactory
- Some formatting improvments
- Improve comments in various places.

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java:
```diff
@@ -40,12 +40,16 @@
 import static org.opentripplanner.util.ElevationUtils.computeEllipsoidToGeoidDifference;
 
 /**
+ * THIS CLASS IS MULTI-THEARED
```
haha typo. Of course it's practically on the first line 😜

landonreed commented on src/main/java/org/opentripplanner/graph_builder/module/ned/ElevationModule.java:
```diff
@@ -229,36 +238,37 @@ public void buildGraph(Graph graph, GraphBuilderModuleSummary graphBuilderModule
-         * Interpolator2D instance.
+         * For unknown reasons, the interpolation of heights at coordinates is a synchronized method in the commonly
+         * used Interpolator2D class. Therefore, it is critical to use a dedicated Coverage instance for each thread to
+         * avoid other threads waiting for a lock to be released on the Coverage instance. This concurrent HashMap will
```
There is no longer a Concurrent HashMap, right?

evansiroky committed:
Improve some comments

evansiroky committed:
Refactor coverage creation to hopefully be thread-safe?

evansiroky committed:
Allow setting a custom amount of parallelism for the elevation module

evansiroky committed:
Only use multithreaded elevation items when multithreading

evansiroky committed:
Make the default amount of processors for the ElevationModule be 1

evansiroky committed:
Improve comments and docs

evansiroky committed:
Correct default elevationModuleParallelism value in build parameter table

evansiroky commented:
Just added some more code that turns off this multi-threading stuff by default and only activates it when a user sets the `elevationModuleParallelism` in build-config.json to a value greater than 1.

evansiroky committed:
Add a GraphWide annotation for failed elevation tile downloads

evansiroky committed:
Various improvements to elevation module

- Add comment about validity of cached elevations based on input data
- Go back to failing hard when downloading NED tiles except for when a tile is missing in the s3 bucket
- Refactor so that Geotiff files are only loaded into GridCoverage2D instances once, but use multiple Interpolator2D instances

binh-dam-ibigroup reviewed (approved):
Amazing.

evansiroky committed:
Update documentation about elevation cache validity

evansiroky committed:
Add EllipsoidToGeoidDifference for known elevations in OSM

evansiroky committed:
Add various improvements

- Make NEDGridCoverageFactory return a UnifiedGridCoverage with thread-specific interpolators
- Use thread local variables instead of fork/join pool in ElevationModule
- Make generic multi-threaded parameter for elevation module

landonreed commented:
@evansiroky, nice work! I'm going to fix the conflicts and merge.

landonreed committed:
Merge branch 'ibi-dev' into edge-process-check
